src/nlr/DeepPolyMaxPoolElement.cpp

```diff
@@ -0,0 +1,227 @@
+/*********************                                                        */
+/*! \file DeepPolyMaxPoolElement.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Haoze Andrew Wu
+ ** This file is part of the Marabou project.
+ ** Copyright (c) 2017-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved. See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** [[ Add lengthier description here ]]
+
+**/
+
+#include "DeepPolyMaxPoolElement.h"
+#include "FloatUtils.h"
+
+namespace NLR {
+
+DeepPolyMaxPoolElement::DeepPolyMaxPoolElement( Layer *layer )
+{
+    _layer = layer;
+    _size = layer->getSize();
+    _layerIndex = layer->getLayerIndex();
+}
+
+DeepPolyMaxPoolElement::~DeepPolyMaxPoolElement()
+{
+    freeMemoryIfNeeded();
+}
+
+void DeepPolyMaxPoolElement::execute( const Map<unsigned, DeepPolyElement *>
+                               &deepPolyElementsBefore )
+{
+    log( "Executing..." );
+    ASSERT( hasPredecessor() );
+    allocateMemory();
+
```

Comment:
If I understood correctly, for the phase fixed case, max is just the identity function; and for the phase not fixed case, the symbolic ub is concretized, and the symbolic lb is the identity for the element with the largest lb?

We can probably use a symbolic expression for the UB as well. There's the work Matan Ostrovsky is doing, and there's also a suggestion the Ehler's Planet paper from 2017.

Reply:
>> for the phase fixed case, max is just the identity function; and for the phase not fixed case, the symbolic ub is concretized, and the symbolic lb is the identity for the element with the largest lb?
<< Yes, that's the formulation in the original paper. 

I'll check with Matan about it. Thanks for the pointer!